### DISCUSSION TITLE
How create INSERT FROM SELECT query with FOR UPDATE hint?

### DISCUSSION THREAD

**Discussion by danillykk** [linq2db/linq2db#4054] [Q&A] [answered]
I'm migrating from BLToolkit to Linq2db and trying to rebuild the InsertIfNotExist query because Linq2db builds it differently.

There is the following table
```
[Table("TestTable")]
public class TestTable
{
	[PrimaryKey]
	public int Id { get; set; }

	[Column("OtherNaming"), NotNull]
	public string Data { get; set; }
		
	[Column("timestampUpdated"), DataType(DataType.Timestamp)]
	public DateTime TimestampUpdated { get; set; }
}
```

I want to make the following request (as with BLToolkit)
```
// Declare block
... 
INSERT INTO `TestTable`
(
	`Id`,
	`OtherNaming`,
	`timestampUpdated`
)
SELECT
	1 as `c1`,
	'Something data' as `c2`,
	@p1 as `c3`
FROM
	(
		SELECT
			Count(*) as `c4`
		FROM
			`TestTable` `a`
		WHERE
			`a`.`Id` = 1
	) `item`
WHERE
	`item`.`c4` = 0 AND TRUE FOR UPDATE
```
a fake field appeared to add FOR UPDATE keyword in BLToolkit
```
[SqlProperty("TRUE FOR UPDATE", ServerSideOnly = true)]
private static bool ForUpdate => true;
```
but Linq2db builds its query differently.
```
...
	) `item`
WHERE
	Count(*) = 0 AND
	TRUE FOR UPDATE = 1 AND
	TRUE FOR UPDATE IS NOT NULL AND
	`a`.`Id` = 1
```

I'm trying to build it on Linq2db
```
...
private readonly ParameterExpression parameterIn = Expression.Parameter(typeof(CountHolder), "input");
private class CountHolder
{
	public int Count { get; set; }
}
...

var query = table.Where(a => a.Id == 1);
var insert = () => new TestTable
	{
		Id = 1,
		Data = "Something data",
		TimestampUpdated = DateTime.UtcNow
	});
var existenceCheckQuery = query.GroupBy(item => new { }).Select(item => new CountHolder {Count = item.Count()});
var commonQuery = existenceCheckQuery.AsSubQuery().Where(item => item.Count == 0).AsMySql().ForUpdateHint();
var setterExpression = Expression.Lambda<Func<CountHolder, T>>(insertSetter.Body, parameterIn);
var insertedCount = commonQuery.Insert(DataConnection.GetTable<T>(), setterExpression);

```
The problem is that the FOR UPDATE keyword doesn't appear at the end of the query.

I also tried to solve the question with the field
```
[Sql.Property(ServerSideOnly = true, ExpectExpression = true, Expression = "TRUE FOR UPDATE", CanBeNull = false, IsPredicate = true)]
private static bool ForUpdate => true;
```
with the following change
```
...
var commonQuery = existenceCheckQuery.AsSubQuery().Where(item => item.Count == 0 && ForUpdate)
...
```
but I get an unnecessary comparison at the end of the query
```
... AND TRUE FOR UPDATE = 1
```

I have no idea how to keep the old behavior

**Comment by MaceWindu** [comment]
- you can use existing hint helpers, e.g. check samples here https://github.com/linq2db/linq2db/blob/master/Tests/Linq/Extensions/PostgreSQLTests.generated.cs
- to avoid explicit bool conversion using comparison, your expression should have `IsPredicate = true`

**Comment by danillykk** [comment]
As I wrote above
1. `.AsMySql().ForUpdateHint()` - this construct doesn't work for the current scenario. But if you do `commonQuery.AsSubQuery().AsMySql().ForUpdateHint()` - the keyword will appear, but it will be inside the subquery
2. `IsPredicate = true` for some reason doesn't cancel the check

**Comment by MaceWindu** [comment]
IsPredicate probably ignored for Sql.Property - you should try Sql.Expression instead.

@igor-tkachev, FYI

**Reply by MaceWindu** [reply]
#4055

**Reply by danillykk** [reply]
```
[Sql.Expression("TRUE FOR UPDATE", IsPredicate = true, CanBeNull = false)]
private static bool ForUpdate => true;
```
this construction doesn't add a hint at the end of the query

**Reply by MaceWindu** [reply, marked as answer]
use this, without serversideonly it is evaluated on client and removed
```cs
[Sql.Expression("TRUE FOR UPDATE", ServerSideOnly = true, CanBeNull = false, IsPredicate = true)]
```

**Reply by danillykk** [reply]
it works, thanks!
